site/content/entry/accurate-audio-emulation.md

```diff
@@ -0,0 +1,199 @@
+
+Nothing.
+
+Even worse, the games exhibited the same behavior from before HLE audio was added to Citra, implying that the games weren't detecting that anything was working at all.
+After spending a few weeks fiddling around with the code, and trying to rule out that it wasn't playing audio because of some bug in the existing code, [wwylele](https://github.com/wwylele) reaffirmed what he always knew.
+When it comes to writing an LLE DSP emulator, it's all or nothing.
+Either all of your code works, or absolutely nothing happens.
+Time to work on DMA.
+
+{put the pdf of his research notes here}
+
+## Burn Out Strikes
+
+Over the next month and into September, [wwylele](https://github.com/wwylele) took shots in the dark and tried to implement DMA.
+He started first with a simple approach to copy data to and from the specific memory regions.
+DMA allows the DSP to copy data quickly both to the 3DS RAM and from the 3DS RAM, making it well suited for uploading audio samples and writing audio output.
+With the simple approach in place, data was being transferred into the emulated DSP, but the output was always zero.
+Debug the code as he might, it was just too frustrating to figure out what was wrong.
+Is there an error in the interpreter? MMIO? Audio Output? Maybe it's actually in DMA?
+At this point, the stress was too much to handle.
+The fear of failure after spending so much time and effort on this project really sunk in, and [wwylele](https://github.com/wwylele) just stopped working on Teakra.
+By the middle of September, [wwylele](https://github.com/wwylele) had just had enough, and decided to work on something else in the mean time.
+
+{pict of convo about dsp status}
+
+It turns out developers are still just regular human beings.
+
+## Success... Finally!
+
+The end of November rolls around with no news about DSP LLE, and [wwylele](https://github.com/wwylele) is working tirelessly to fix many other core issues, including [fixing issues with how citra handled memory that prevented many cheat codes from working properly](https://github.com/citra-emu/citra/pull/4392) (but more on that in another blog post!)
+Out of nowhere, there's new progress on DMA.
+[wwylele](https://github.com/wwylele) worked out some extra details about the different patterns that the DMA can use to copy memory and things started taking off from there.
+The next day, he noticed an [oversight](https://github.com/wwylele/teakra/commit/d22c2d99d3b0858eb190bf9298815fcac237793a#diff-fe5ba875bf987b9df6685326294fe86fR13) in his simple DMA approach, which ended up causing the DMA to copy data to the *program* region in memory instead of the data region!
+Now that data was going to the right spot, it was much easier to debug.
+
+After some more reverse engineering and hardware testing, wwylele had a good idea of how to recreate DMA completely now, and tried it out on a custom application that produces just a simple sine wave.
```

Comment:
The "custom application" is actually an example program in devkitPro ("audio/streaming")

Reply:
I surprisingly knew that, but explaining that its a sample homebrew is probably too much effort for little pay off at this point in the article